pkg/executor/join/hash_join_v2.go

```diff
@@ -0,0 +1,870 @@
+	}
+	return nil
+}
+
+func (fetcher *ProbeSideTupleFetcherV2) shouldLimitProbeFetchSize() bool {
+	if fetcher.JoinType == plannercore.LeftOuterJoin && fetcher.RightAsBuildSide {
+		return true
+	}
+	if fetcher.JoinType == plannercore.RightOuterJoin && !fetcher.RightAsBuildSide {
+		return true
+	}
+	return false
+}
+
+func (w *BuildWorkerV2) splitPartitionAndAppendToRowTable(typeCtx types.Context, srcChkCh chan *chunk.Chunk) (err error) {
+	cost := int64(0)
+	defer func() {
+		if w.HashJoinCtx.stats != nil {
+			atomic.AddInt64(&w.HashJoinCtx.stats.partitionData, cost)
+			setMaxValue(&w.HashJoinCtx.stats.maxPartitionData, cost)
+		}
+	}()
+	partitionNumber := w.HashJoinCtx.PartitionNumber
+	hashJoinCtx := w.HashJoinCtx
+
+	builder := createRowTableBuilder(w.BuildKeyColIdx, hashJoinCtx.BuildKeyTypes, partitionNumber, w.HasNullableKey, hashJoinCtx.BuildFilter != nil, hashJoinCtx.needScanRowTableAfterProbeDone)
+
+	for chk := range srcChkCh {
+		start := time.Now()
+		err = builder.processOneChunk(chk, typeCtx, w.HashJoinCtx, int(w.WorkerID))
+		failpoint.Inject("splitPartitionPanic", nil)
+		cost += int64(time.Since(start))
+		if err != nil {
+			return err
+		}
+	}
+	start := time.Now()
+	builder.appendRemainingRowLocations(int(w.WorkerID), w.HashJoinCtx.hashTableContext)
+	cost += int64(time.Since(start))
+	return nil
+}
+
+func (e *HashJoinV2Exec) canSkipProbeIfHashTableIsEmpty() bool {
+	switch e.JoinType {
+	case plannercore.InnerJoin:
+		return true
+	case plannercore.LeftOuterJoin:
+		return !e.RightAsBuildSide
+	case plannercore.RightOuterJoin:
+		return e.RightAsBuildSide
+	case plannercore.SemiJoin:
+		return e.RightAsBuildSide
+	default:
+		return false
+	}
+}
+
+func (e *HashJoinV2Exec) initializeForProbe() {
+	e.ProbeSideTupleFetcher.HashJoinCtxV2 = e.HashJoinCtxV2
+	// e.joinResultCh is for transmitting the join result chunks to the main
+	// thread.
+	e.joinResultCh = make(chan *hashjoinWorkerResult, e.Concurrency+1)
+	e.ProbeSideTupleFetcher.initializeForProbeBase(e.Concurrency, e.joinResultCh)
+	e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty = e.canSkipProbeIfHashTableIsEmpty()
+
+	for i := uint(0); i < e.Concurrency; i++ {
+		e.ProbeWorkers[i].initializeForProbe(e.ProbeSideTupleFetcher.probeChkResourceCh, e.ProbeSideTupleFetcher.probeResultChs[i], e)
+	}
+}
+
+func (e *HashJoinV2Exec) fetchAndProbeHashTable(ctx context.Context) {
+	e.initializeForProbe()
+	e.workerWg.RunWithRecover(func() {
+		defer trace.StartRegion(ctx, "HashJoinProbeSideFetcher").End()
+		e.ProbeSideTupleFetcher.fetchProbeSideChunks(ctx, e.MaxChunkSize(), func() bool {
+			return e.ProbeSideTupleFetcher.hashTableContext.hashTable.isHashTableEmpty()
+		}, e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty, e.ProbeSideTupleFetcher.needScanRowTableAfterProbeDone,
+			e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize(), &e.ProbeSideTupleFetcher.hashJoinCtxBase)
+	}, e.ProbeSideTupleFetcher.handleProbeSideFetcherPanic)
```

Comment:
```suggestion
	fetchProbeSideChunksFunc := func() {
		defer trace.StartRegion(ctx, "HashJoinProbeSideFetcher").End()

		e.ProbeSideTupleFetcher.fetchProbeSideChunks(
			ctx,
			e.MaxChunkSize(),
			func() bool {
				return e.ProbeSideTupleFetcher.hashTableContext.hashTable.isHashTableEmpty()
			},
			e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty,
			e.ProbeSideTupleFetcher.needScanRowTableAfterProbeDone,
			e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize(),
			&e.ProbeSideTupleFetcher.hashJoinCtxBase,
		)
	}

	e.workerWg.RunWithRecover(fetchProbeSideChunksFunc, e.ProbeSideTupleFetcher.handleProbeSideFetcherPanic)

```
Maybe this is more clear